cli_tools/osconfig_agent/patch_window.go

```diff
@@ -15,26 +15,306 @@
+	a.mx.Lock()
+	a.windows[name] = w
+	a.mx.Unlock()
+}
+
+type patchPolicy struct {
+	*osconfigpb.LookupConfigsResponse_EffectivePatchPolicy
+}
+
+// MarshalJSON marchals a patchPolicy using jsonpb.
+func (p *patchPolicy) MarshalJSON() ([]byte, error) {
+	m := jsonpb.Marshaler{}
+	s, err := m.MarshalToString(p)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(s), nil
+}
+
+// UnmarshalJSON unmarshals apatchPolicy using jsonpb.
+func (p *patchPolicy) UnmarshalJSON(b []byte) error {
+	return jsonpb.UnmarshalString(string(b), p)
+}
+
 type patchWindow struct {
-	start, end time.Time
+	Name   string
+	Policy *patchPolicy
+
+	Start, End time.Time
+
+	t      *time.Timer
+	cancel chan struct{}
+	mx     sync.RWMutex
+}
+
+func (w *patchWindow) String() string {
+	return w.Name
+}
+
+func newPatchWindow(pp *osconfigpb.LookupConfigsResponse_EffectivePatchPolicy) (*patchWindow, error) {
+	start, end, err := nextWindow(time.Now().UTC(), pp.PatchWindow, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &patchWindow{
+		Name:   pp.GetFullName(),
+		Policy: &patchPolicy{pp},
+		Start:  start,
+		End:    end,
+	}, nil
 }
 
 func (w *patchWindow) in() bool {
+	select {
+	case <-w.cancel:
+		return false
+	default:
+	}
+
 	now := time.Now()
-	return now.After(w.start) && now.Before(w.end)
+	return now.After(w.Start) && now.Before(w.End)
+}
+
+// update updates a patchWindow with a PatchPolicy.
+func (w *patchWindow) update(p *osconfigpb.LookupConfigsResponse_EffectivePatchPolicy) {
+	fmt.Println("DEBUG: update", w.Name)
+	start, end, err := nextWindow(time.Now().UTC(), p.PatchWindow, 0)
+	if err != nil {
+		return
+	}
+
+	if start != w.Start {
+		fmt.Println("DEBUG: patchWindow start changed, need to reregister:", w.Name)
+		w.t.Stop()
+		// Cancel any potential current run if we are no longer in a patch window.
+		if w.in() {
+			fmt.Println("DEBUG: patchWindow running, need to cancel:", w.Name)
+			close(w.cancel)
+		}
+
+		defer w.register()
+	}
+
+	w.mx.Lock()
+	w.Start, w.End = start, end
+	w.Policy = &patchPolicy{p}
+	w.mx.Unlock()
+
+	fmt.Println("DEBUG: update end", w.Name)
+}
+
+// register registers a patch window as active, this will clobber any existing
+// window with the same name.
+func (w *patchWindow) register() {
+	fmt.Println("DEBUG: register", w.Name)
+	w.mx.Lock()
+	defer w.mx.Unlock()
+	fmt.Println("DEBUG: register got lock", w.Name)
+	defer func() { fmt.Println("DEBUG: register unlock", w.Name) }()
+
+	aw.add(w.Name, w)
+
+	// Create the Timer that will kick off the patch process.
+	// If we happen to be in the patch window now this will start imediately.
+	w.cancel = make(chan struct{})
+	w.t = time.AfterFunc(w.Start.Sub(time.Now()), func() { rc <- w })
+	fmt.Println(w.Name, "register done")
+}
+
+// deregister stops a patch windows timer and removes it from activeWindows
+// but does not cancel any current runs.
+func (w *patchWindow) deregister() {
+	fmt.Println(w.Name, "deregister")
+	w.t.Stop()
+	aw.delete(w.Name)
```

Comment:
It seems unconventional that an item in a list, deletes itself from that list. Maybe this is something that you're going to change or maybe there's a good reason for it. But it's different from what I'd usually expect to see and can be confusing when following the code.